zkwrapper/zkwrapper.cc

```diff
@@ -0,0 +1,534 @@
+	{-1, "ZSYSTEMERROR"},
+	{-2, "ZRUNTIMEINCONSISTENCY"},
+	{-3, "ZDATAINCONSISTENCY"},
+	{-4, "ZCONNECTIONLOSS"},
+	{-5, "ZMARSHALLINGERROR"},
+	{-6, "ZUNIMPLEMENTED"},
+	{-7, "ZOPERATIONTIMEOUT"},
+	{-8, "ZBADARGUMENTS"},
+	{-9, "ZINVALIDSTATE"},
+	{-100, "ZAPIERROR"},
+	{-101, "ZNONODE"},
+	{-102, "ZNOAUTH"},
+	{-103, "ZBADVERSION"},
+	{-108, "ZNOCHILDRENFOREPHEMERALS"},
+	{-110, "ZNODEEXISTS"},
+	{-111, "ZNOTEMPTY"},
+	{-112, "ZSESSIONEXPIRED"},
+	{-113, "ZINVALIDCALLBACK"},
+	{-114, "ZINVALIDACL"},
+	{-115, "ZAUTHFAILED"},
+	{-116, "ZCLOSING"},
+	{-117, "ZNOTHING"},
+	{-118, "ZSESSIONMOVED"},
+	{-120, "ZNEWCONFIGNOQUORUM"},
+	{-121, "ZRECONFIGINPROGRESS"},
+	{-999, "ZKWRAPPERDEFAULTERROR"},
+};
+
+/**
+ * TODO
+ *
+ * @param zzh zookeeper handle
+ * @param type type of event
+ * @param state state of the event
+ * @param path path to the watcher node
+ * @param watcherCtx the state of the watcher
+ */
+void watcher(zhandle_t *zzh,
+		int type,
+		int state,
+		const char *path,
+		void *watcherCtx) {
+	LOG(INFO) << "[Global watcher] Watcher triggered on path '" << path << "'"
+		;
+	char health[] = "/health/datanode_";
+	if (type == ZOO_SESSION_EVENT) {
+		if (state == ZOO_CONNECTED_STATE) {
+			return;
+		} else if (state == ZOO_AUTH_FAILED_STATE) {
+			zookeeper_close(zzh);
+			exit(1);
+		} else if (state == ZOO_EXPIRED_SESSION_STATE) {
+			zookeeper_close(zzh);
+			exit(1);
+		}
+	}
+}
+
+
+watcher_fn ZKWrapper::watcher_health_factory(std::string path){
+	class factory_wrapper{
+		public:
+
+			static void watcher_health(zhandle_t *zzh, int type, int state, const char *path, void *watcherCtx) {
+
+				struct String_vector stvector;
+				struct String_vector *vector = &stvector;
+				/* reinstall watcher */
+				int rc = zoo_wget_children(zzh, path, watcher_health, nullptr, vector);
+				LOG(INFO)  <<  "[In watcher_health rc] health:" << rc;
+				int i;
+				std::vector <std::string> children;
+				for (i = 0; i < stvector.count; i++) {
+					children.push_back(stvector.data[i]);
+				}
+
+				if (children.size() == 0){
+					// ZkNnClient::CLASS_NAME is not in scope when put into zkwrapper
+					LOG(INFO) <<  "no childs to retrieve";
+				}
+
+				for (int i = 0; i < children.size(); i++) {
+					LOG(INFO) <<  "[In watcher_health] Attaching child to " << children[i];
+					//ZkClientCommon::HEALTH_BACKSLASH + children[i]).c_str(),
+					int rc = zoo_wget_children(zzh, (path+children[i]).c_str(),
+							ZKWrapper::watcher_health_child, nullptr,
+							vector);
+				}
+			}
+	};
+    return factory_wrapper::watcher_health;	
+} 
+
+/*
+ * Watcher for health child node (/health/datanode_)
+ */
+void ZKWrapper::watcher_health_child(zhandle_t *zzh, int type, int state, const char *path, void *watcherCtx) {
+	LOG(INFO) << CLASS_NAME << "[health child] Watcher triggered on path '" << path;
+	char health[] = "/health/datanode_";
+	LOG(INFO) << CLASS_NAME << "[health child] Receive a heartbeat. A child has been added under path" << path;
```

Comment:
I don't think we can assume a child has been added. It's possible that a child has been removed.